src/validation.cpp

```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
     LimitMempoolSize(mempool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
 }
 
+// Used to avoid mempool polluting consensus critical paths if CCoinsViewMempool
+// were somehow broken and returning the wrong scriptPubKeys
+static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &view, CTxMemPool& pool,
+                 unsigned int flags, bool cacheSigStore, PrecomputedTransactionData& txdata) {
+    AssertLockHeld(cs_main);
+
+    assert(!tx.IsCoinBase());
+    {
+        LOCK(pool.cs);
+        for (const CTxIn& txin : tx.vin) {
+            const Coin& coin = view.AccessCoin(txin.prevout);
+            if (coin.IsSpent()) return false;
+
+            const CTransactionRef& txFrom = pool.get(txin.prevout.hash);
+            if (txFrom) {
```

Comment:
I'd really prefer not to. The unordered_map lookup should essentially be free since anything required to find that transaction should already be very hot in cache, and if its free, why shouldnt we?